pkg/segment/query/processor/queryprocessor.go

```diff
@@ -216,8 +223,77 @@ func (qp *QueryProcessor) GetFullResult() (*structs.PipeSearchResponseOuter, err
 // 3. Read from the update channel and the final result channel.
 //
 // Once the final result is sent, no more updates will be sent.
-func (qp *QueryProcessor) GetStreamedResult(updateChan chan *structs.PipeSearchWSUpdateResponse,
-	completeChan chan *structs.PipeSearchCompleteResponse) {
+func (qp *QueryProcessor) GetStreamedResult(stateChan chan *query.QueryStateChanData) error {
+
+	var finalIQR *iqr.IQR
+	var err error
+	totalRecords := 0
+
+	var iqr *iqr.IQR
+	completeResp := &structs.PipeSearchCompleteResponse{
+		Qtype: qp.queryType.String(),
+	}
+
+	for err != io.EOF {
+		iqr, err = qp.DataProcessor.Fetch()
+		if err != nil && err != io.EOF {
+			return utils.TeeErrorf("GetStreamedResult: failed to fetch iqr, err: %v", err)
+		}
+		if iqr == nil {
+			break
+		}
+		if finalIQR == nil {
+			finalIQR = iqr
+		} else {
+			appendErr := finalIQR.Append(iqr)
+			if appendErr != nil {
+				return utils.TeeErrorf("GetStreamedResult: failed to append iqr to the finalIQR, err: %v", appendErr)
+			}
+		}
+
+		if qp.queryType == structs.RRCCmd && iqr.NumberOfRecords() > 0 {
+			result, wsErr := iqr.AsWSResult(qp.queryType, qp.timeOrdered)
+			if wsErr != nil {
+				return utils.TeeErrorf("GetStreamedResult: failed to get WSResult from iqr, wsErr: %v", err)
+			}
+			totalRecords += len(result.Hits.Hits)
+			stateChan <- &query.QueryStateChanData{
+				StateName:    query.QUERY_UPDATE,
+				UpdateWSResp: result,
+			}
+		}
+	}
+
+	if qp.queryType != structs.RRCCmd {
+		result, err := finalIQR.AsWSResult(qp.queryType, qp.timeOrdered)
+		if err != nil {
+			return utils.TeeErrorf("GetStreamedResult: failed to get WSResult from iqr; err: %v", err)
+		}
+		completeResp.MeasureResults = result.MeasureResults
+		completeResp.MeasureFunctions = result.MeasureFunctions
+		completeResp.GroupByCols = result.GroupByCols
+		completeResp.BucketCount = result.BucketCount
+	}
+
+	progress, err := query.GetProgress(qp.qid)
+	if err != nil {
+		return utils.TeeErrorf("GetStreamedResult: failed to get progress; err: %v", err)
+	}
+
+	relation := "eq"
+	if progress.UnitsSearched < progress.TotalUnits {
+		relation = "gte"
+	}
+
+	completeResp.TotalMatched = utils.HitsCount{Value: uint64(totalRecords), Relation: relation}
```

Comment:
I don't think this is right. When the full search is done you'd be using `eq`, but the query may have finished early (i.e., sent enough results to the UI to mark the query as completed, without actually searching all the segments that might have a match), and in that case the number of records matching the query could be greater than the number of results returned. But if the query didn't exit early, then we can (and should) use `eq`

Reply:
We will only returned the number of records that we computed by the query and not the matched results similar to the older pipeline. I have talked about this with Kunal, that we want to show the final number of records instead of showing the totalMatched of the initial search. 
I will only send `eq` if we have gone through all the segments/blocks which is correct. In case of query exits early we can safely say that results could be more by sending `gte`.
@nkunal can confirm.

Reply:
1. when we are exiting early we must send `gte` . 
2. When we have searched all the segments and we are sending the exact number of records back to the user (for e.g. only 34 records matched out of 5 million) then we should send `eq`
3. When we have searched all the segments and we are sending a subset of records back to the user (for e.g. sending only 500 records although matched count was 100K out of 5 million searched) then we should send `gte`